From df0035a8d8249685a3b41ca498c1190f559710b9 Mon Sep 17 00:00:00 2001
From: Sung Joon Kim <sungkim@amd.com>
Date: Fri, 7 Oct 2022 13:48:56 -0400
Subject: [PATCH 011/161] amd/display: fix for dmub stuttering

(cherry picked from commit 505617021bba482d204bcde6e1a444ac743c77fb)
[Fwd-ported to v6.11]
Signed-off-by: Cristian Ciocaltea <cristian.ciocaltea@collabora.com>
[Fwd-ported to v6.14]
Signed-off-by: Robert Beckett <bob.beckett@collabora.com>
---
 .../gpu/drm/amd/display/amdgpu_dm/amdgpu_dm_backlight.c  | 9 +++++++--
 1 file changed, 7 insertions(+), 2 deletions(-)

diff --git a/drivers/gpu/drm/amd/display/amdgpu_dm/amdgpu_dm_backlight.c b/drivers/gpu/drm/amd/display/amdgpu_dm/amdgpu_dm_backlight.c
index 703d0617d..35378d2de 100644
--- a/drivers/gpu/drm/amd/display/amdgpu_dm/amdgpu_dm_backlight.c
+++ b/drivers/gpu/drm/amd/display/amdgpu_dm/amdgpu_dm_backlight.c
@@ -323,6 +323,7 @@ void amdgpu_dm_backlight_set_level(struct amdgpu_display_manager *dm,
 	struct amdgpu_dm_backlight_caps *caps;
 	struct dc_link *link;
 	u32 brightness = 0;
+	u32 prev_brightness;
 	u32 millipercent = 0;
 	bool rc = false, reallow_idle = false;
 	struct drm_connector *connector;
@@ -345,7 +346,10 @@ void amdgpu_dm_backlight_set_level(struct amdgpu_display_manager *dm,
 	amdgpu_dm_update_backlight_caps(dm, bl_idx);
 	caps = &dm->backlight_caps[bl_idx];
 
+	prev_brightness = dm->brightness[bl_idx];
 	dm->brightness[bl_idx] = user_brightness;
+	dm->actual_brightness[bl_idx] = user_brightness;
+
 	/* update scratch register */
 	if (bl_idx == 0)
 		amdgpu_atombios_scratch_regs_set_backlight_level(dm->adev, dm->brightness[bl_idx]);
@@ -411,8 +415,8 @@ void amdgpu_dm_backlight_set_level(struct amdgpu_display_manager *dm,
 
 	mutex_unlock(&dm->dc_lock);
 
-	if (rc)
-		dm->actual_brightness[bl_idx] = user_brightness;
+	if (!rc)
+		dm->actual_brightness[bl_idx] = prev_brightness;
 }
 EXPORT_IF_KUNIT(amdgpu_dm_backlight_set_level);
 
@@ -541,6 +545,7 @@ amdgpu_dm_register_backlight_device(struct amdgpu_dm_connector *aconnector)
 		backlight_device_register(bl_name, aconnector->base.kdev, dm,
 					  &amdgpu_dm_backlight_ops, &props);
 	dm->brightness[aconnector->bl_idx] = props.brightness;
+	dm->actual_brightness[aconnector->bl_idx] = props.brightness;
 
 	if (IS_ERR(dm->backlight_dev[aconnector->bl_idx])) {
 		drm_err(drm, "DM: Backlight registration failed!\n");
-- 
2.56.0

